File: server/src/questions/question.model.ts
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Type } from 'class-transformer';
import mongoose from 'mongoose';
import { Document } from 'mongoose';
import { Choice, ChoiceDocument, ChoiceSchema } from 'src/choices/choices.model';
import { Game } from 'src/games/game.model';


enum QuestionType {
    Quiz=10,
    TrueFalse=20,
    TypeAnswer=30,
    Puzzle=40
}
export type QuestionDocument = Question & Document;

@Schema({
    toJSON: {
        getters: true,
        virtuals: true,
      }, 
      toObject:{
          getters: true,
          virtuals: true,
      }, 
  })
export class Question {
  @Prop({ type: mongoose.Schema.Types.ObjectId, ref: 'Game',  required: true })
  @Type(() => Game)
  gameId: Game;
 

  @Prop({ required: true })
  questionText: string;

  @Prop()
  questionPhoto: string;

  @Prop({ required: true, default:10 })
  questionPoint: number;

  @Prop({ required: true, default:20 })
  timeLimit: number;

  @Prop({ default: false })
  isMultiSelect: boolean;

  @Prop({
    enum: QuestionType, 
    required: true, 
    default:QuestionType.Quiz})
  type: number;

  @Type(() => Choice)
  choices: Choice[];

}
const QuestionSchema = SchemaFactory.createForClass(Question);

QuestionSchema.virtual('choices', {
    ref: 'Choice',
    localField: '_id',
    foreignField: 'questionId',
});


export { QuestionSchema };

File: client/lib/models/question_model.dart
import 'package:kahootlike_game/models/choice_model.dart';

class Question {
  String? id;
  String? gameId;
  String? questionText;
  String? questionPhoto;
  int? questionPoint;
  int? timeLimit;
  bool? isMultiSelect;
  int? type;
  List<Choice>? choices;

  Question(
      {this.id,
      this.gameId,
      this.questionText,
      this.questionPhoto,
      this.questionPoint,
      this.timeLimit,
      this.isMultiSelect,
      this.type,
      this.choices});

  Question.fromJson(Map<String, dynamic> json) {
    id = json['_id'];
    gameId = json['gameId'];
    questionText = json['questionText'];
    questionPhoto = json['questionPhoto'];
    questionPoint = json['questionPoint'];
    timeLimit = json['timeLimit'];
    isMultiSelect = json['isMultiSelect'];
    type = json['type'];
    if (json['choices'] != null) {
      choices = <Choice>[];
      json['choices'].forEach((v) {
        choices!.add(new Choice.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.id;
    data['gameId'] = this.gameId;
    data['questionText'] = this.questionText;
    data['questionPhoto'] = this.questionPhoto;
    data['questionPoint'] = this.questionPoint;
    data['timeLimit'] = this.timeLimit;
    data['isMultiSelect'] = this.isMultiSelect;
    data['type'] = this.type;
    if (this.choices != null) {
      data['choices'] = this.choices!.map((v) => v.toJson()).toList();
    }
    data.removeWhere((key, value) => key == null || value == null);
    return data;
  }
}
